finality-aleph/src/data_io/data_store.rs

```diff
@@ -0,0 +1,636 @@
+                    self.on_block_finalized((block.header.hash(), *block.header.number()).into());
+                }
+                _ = &mut maintenance_clock => {
+                    self.run_maintenance();
+                    maintenance_clock = Delay::new(self.config.periodic_maintenance_interval);
+                }
+                _ = &mut exit => {
+                    debug!(target: "afa", "Data Store task received exit signal. Terminating.");
+                    break;
+                }
+            }
+        }
+    }
+    // Updates our highest known and highest finalized block info directly from the client.
+    fn update_highest_blocks(&mut self) {
+        let client_info = self.client.info();
+        if let Some(header) = self
+            .client
+            .header(BlockId::Hash(client_info.best_hash))
+            .expect("client must answer a query")
+        {
+            self.on_block_imported(header);
+        } else {
+            warn!(target: "afa", "Client info has best_block at {:?} but the block not available in the header backend.", client_info.best_number);
+        }
+
+        if self.highest_finalized_num < client_info.finalized_number {
+            self.on_block_finalized(
+                (client_info.finalized_hash, client_info.finalized_number).into(),
+            );
+        }
+    }
+
+    fn run_maintenance(&mut self) {
+        self.update_highest_blocks();
+
+        let proposals_with_timestamps: Vec<_> = self
+            .pending_proposals
+            .iter()
+            .map(|(proposal, info)| (proposal.clone(), info.first_occurence))
+            .collect();
+        if !proposals_with_timestamps.is_empty() {
+            info!(target: "afa", "Data Store maintenance. Awaiting {:?} proposals.",proposals_with_timestamps.len());
+            if proposals_with_timestamps.len() <= 5 {
+                info!(target: "afa", "Pending proposals: {:?}", proposals_with_timestamps);
+            } else {
+                info!(target: "afa", "Pending proposals (showing 5 initial only): {:?}", &proposals_with_timestamps[..5]);
+            }
+        }
+
+        let now = time::SystemTime::now();
+        for (proposal, first_occurence) in proposals_with_timestamps {
+            if !self.bump_proposal(&proposal) {
+                if let Ok(time_waiting) = now.duration_since(first_occurence) {
+                    if time_waiting >= self.config.request_block_after {
+                        let block = proposal.top_block();
+                        if !self
+                            .proposal_status_provider
+                            .lock()
+                            .is_block_imported(&block)
+                            && !self.requested_blocks_cache.contains(&block)
+                        {
+                            // We remember the block to not request it again.
```

Comment:
I said that for the local tests adding this logic didn't help. I have hopes that it might be helpful in the real network. All in all, I think there is no harm in having it here.

Reply:
More code is always harm, more place to insert bugs. I wish I was joking. Dx